### PR TITLE
run prune logic in each controllers

### PR DESCRIPTION
**Pull request by ldpliu**
## Related issue(s)
https://issues.redhat.com/browse/ACM-15400
Fixes #

## Tests
* [x] Unit/function tests have been added and incorporated into `make unit-tests`.
* [x] Integration tests have been added and incorporated into `make integration-test`.
* [x] E2E tests have been added and incorporated into `make e2e-test-all`.
* [x] List other manual tests you have done.

### CODE REVIEW

File: operator/pkg/config/transport_config.go [hunk 1]
```diff
@@ -43,10 +43,12 @@ var (
 )
 
 func SetTransporterConn(conn *transport.KafkaConfig) {
+	log.Info("Set Transporter Conn")
```
Comment (clyang82) [inline operator/pkg/config/transport_config.go:46]: keep it as the same as others to use Debug.

File: operator/pkg/config/transport_config.go [hunk 2]
```diff
@@ -43,10 +43,12 @@ var (
...
 	transporterConn = conn
 }
 
 func GetTransporterConn() *transport.KafkaConfig {
+	log.Debugf("Set Transporter Conn: %v", transporterConn != nil)
```
Comment (clyang82) [inline operator/pkg/config/transport_config.go:51]: "Get Transporter Conn"

File: operator/pkg/controllers/acm/resources.go [hunk 3]
```diff
@@ -51,6 +51,16 @@ type ACMResourceController struct {
 	Resources map[string]bool
 }
 
+type ACMMetaController struct{}
```
Comment (clyang82) [inline operator/pkg/controllers/acm/resources.go:54]: why do you want to call `ACMMetaController`? a metacontroller acts as a "controller of controllers". We should only have one meta controller which is used to start all other controllers.

**Review by clyang82 — CHANGES_REQUESTED** [review]

File: operator/pkg/config/transport_config.go (hunk 1 above)
Comment (ldpliu) [reply operator/pkg/config/transport_config.go]: done

File: operator/pkg/config/transport_config.go (hunk 2 above)
Comment (ldpliu) [reply operator/pkg/config/transport_config.go]: done

File: operator/pkg/controllers/acm/resources.go (hunk 3 above)
Comment (ldpliu) [reply operator/pkg/controllers/acm/resources.go]: changed

File: operator/pkg/config/addon_config.go [hunk 4]
```diff
@@ -34,6 +34,8 @@ import (
 
 const SignerName = "open-cluster-management.io/globalhub-signer"
 
+var isAgentRemoved bool
```
[the remark this answers is not in the document]
Comment (ldpliu) [reply operator/pkg/config/addon_config.go]: changed the name, it meas if all the addons deleted.

File: operator/pkg/controllers/agent/addon_manager.go [hunk 5]
```diff
@@ -42,6 +42,16 @@ var FS embed.FS
 
 var started bool
 
+type AddonManagerMetaController struct{}
```
[the remark this answers is not in the document]
Comment (ldpliu) [reply operator/pkg/controllers/agent/addon_manager.go]: changed the logic

File: operator/pkg/controllers/transporter/protocol/strimzi_kafka_controller.go [hunk 6]
```diff
@@ -37,28 +47,49 @@ type KafkaStatus struct {
...
 	}
+	if mgh.DeletionTimestamp != nil {
+		if !config.GetGlobalhubAgentRemoved() {
+			return ctrl.Result{RequeueAfter: 5 * time.Second}, nil
+		}
+		return r.pruneStrimziResources(ctx)
+	}
+	isResourceRemoved = false
+	if !mgh.Spec.EnableMetrics {
+		err = operatorutils.PruneMetricsResources(ctx, r.c,
```
Comment (clyang82) [inline operator/pkg/controllers/transporter/protocol/strimzi_kafka_controller.go:82]: I think it should be included in `mgh.DeletionTimestamp` judgement.

**Pull request merged**